Subject: [PATCH] Enforce the custom blocklist in SDK37 package scans

The LDU scan branch was removed in SDK37. Check the same custom list used
by InstallPackageHelper before scanPackageOnly can create a ScanResult.
Keep the explicit shell bypass effective for install scans too; it defaults
to false at each system_server start. RDU state is not a custom-list gate.

diff --git a/smali_classes2/com/android/server/pm/ScanPackageUtils.smali b/smali_classes2/com/android/server/pm/ScanPackageUtils.smali
--- a/smali_classes2/com/android/server/pm/ScanPackageUtils.smali
+++ b/smali_classes2/com/android/server/pm/ScanPackageUtils.smali
@@ -1164,6 +1164,12 @@
 
     iget-object v6, v1, Lcom/android/server/pm/ScanRequest;->mParsedPackage:Lcom/android/internal/pm/parsing/pkg/ParsedPackage;
 
+    invoke-interface {v6}, Lcom/android/internal/pm/parsing/pkg/ParsedPackage;->getPackageName()Ljava/lang/String;
+
+    move-result-object v3
+
+    invoke-static {v3}, Lcom/samsung/android/server/pm/install/PackageBlockListPolicy;->checkPackageForScan(Ljava/lang/String;)V
+
     iget-object v0, v1, Lcom/android/server/pm/ScanRequest;->mPkgSetting:Lcom/android/server/pm/PackageSetting;
 
     iget-object v7, v1, Lcom/android/server/pm/ScanRequest;->mDisabledPkgSetting:Lcom/android/server/pm/PackageSetting;
diff --git a/smali_classes2/com/samsung/android/server/pm/install/PackageBlockListPolicy.smali b/smali_classes2/com/samsung/android/server/pm/install/PackageBlockListPolicy.smali
--- a/smali_classes2/com/samsung/android/server/pm/install/PackageBlockListPolicy.smali
+++ b/smali_classes2/com/samsung/android/server/pm/install/PackageBlockListPolicy.smali
@@ -27,3 +27,53 @@
 
     return-void
 .end method
+
+.method public static checkPackageForScan(Ljava/lang/String;)V
+    .locals 3
+
+    sget-boolean v0, Lcom/samsung/android/server/pm/install/PackageBlockListPolicy;->sBypassBlocklist:Z
+
+    if-nez v0, :allowed
+
+    sget-object v0, Lcom/samsung/android/server/pm/install/PackageBlockListPolicy;->sBlocklist:Ljava/util/HashSet;
+
+    if-nez v0, :loaded
+
+    const-string v0, "/system/etc/unica_blocklist.xml"
+
+    invoke-static {v0}, Lcom/samsung/android/server/pm/install/PmConfigParser;->parsePackages(Ljava/lang/String;)Ljava/util/List;
+
+    move-result-object v0
+
+    new-instance v1, Ljava/util/HashSet;
+
+    invoke-direct {v1, v0}, Ljava/util/HashSet;-><init>(Ljava/util/Collection;)V
+
+    sput-object v1, Lcom/samsung/android/server/pm/install/PackageBlockListPolicy;->sBlocklist:Ljava/util/HashSet;
+
+    :loaded
+    sget-object v0, Lcom/samsung/android/server/pm/install/PackageBlockListPolicy;->sBlocklist:Ljava/util/HashSet;
+
+    invoke-virtual {v0, p0}, Ljava/util/HashSet;->contains(Ljava/lang/Object;)Z
+
+    move-result v0
+
+    if-eqz v0, :allowed
+
+    new-instance v0, Lcom/android/server/pm/PackageManagerException;
+
+    const/16 v1, -0x6e
+
+    const-string v2, "Package blocked by /system/etc/unica_blocklist.xml: "
+
+    invoke-virtual {v2, p0}, Ljava/lang/String;->concat(Ljava/lang/String;)Ljava/lang/String;
+
+    move-result-object v2
+
+    invoke-direct {v0, v1, v2}, Lcom/android/server/pm/PackageManagerException;-><init>(ILjava/lang/String;)V
+
+    throw v0
+
+    :allowed
+    return-void
+.end method
